### PR TITLE
fix: compatibility with pydantic v2

### PR DESCRIPTION
**Pull request by dagondagon666**
**What this PR does / why we need it**:
- Follows Optional fields guidance for pydantic v2 [ref](https://docs.pydantic.dev/latest/migration/#required-optional-and-nullable-fields)
- Overwrites protected_namespace in pydantic v2 to get rid of warning logs about namespace conflicts in pydantic [ref](https://docs.pydantic.dev/latest/api/config/#protected-namespaces). An example is shown as follows
```
E   UserWarning: Field "model_name" has conflict with protected namespace "model_".
E
E   You may be able to resolve this warning by setting `model_config['protected_namespaces'] = ()`.
```

**Type of changes**
Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.
**No image version increase**

**Checklist**:

- [ ] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [ ] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Review by terrytangyuan — COMMENTED** [review]
Could you add a test for this?

File: python/kserve/kserve/protocol/rest/v2_datamodels.py [hunk 1]
```diff
@@ -14,7 +14,7 @@
 from typing import Optional, List, Union, Dict
 
 import orjson
-from pydantic import BaseModel
+from pydantic import BaseModel, ConfigDict
```
Comment (terrytangyuan) [inline python/kserve/kserve/protocol/rest/v2_datamodels.py:17]: Should we surround the imports with try-catch in for backwards compatibility purposes?

File: python/kserve/kserve/protocol/rest/v2_datamodels.py [hunk 2]
```diff
@@ -229,10 +229,11 @@ class InferenceResponse(BaseModel):
         }
     """
     model_name: str
-    model_version: Optional[str]
+    model_version: Optional[str] = None
     id: str
-    parameters: Optional[Parameters]
+    parameters: Optional[Parameters] = None
     outputs: List[ResponseOutput]
+    model_config = ConfigDict(protected_namespaces='')
```
Comment (yuzisun) [inline python/kserve/kserve/protocol/rest/v2_datamodels.py:236]: Why do we need the `model_config` ?

Comment (timothyjlaurent) [reply python/kserve/kserve/protocol/rest/v2_datamodels.py:236]: This is the new syntax for pydantic 2 that's analogous to the nested Config class in pydantic 1.

My guess is that Pydantic2 uses "model_" a protected namespace which we need for `model_version`

File: python/kserve/kserve/protocol/rest/v2_datamodels.py [hunk 3]
```diff
@@ -229,10 +229,11 @@ class InferenceResponse(BaseModel):
...
 
     class Config:
```
Comment (timothyjlaurent) [inline python/kserve/kserve/protocol/rest/v2_datamodels.py:238]: This is pydantic 1 syntax.

We should move the `schema_extra` into the `json_schema_extra` field in the `model_config` above.

https://docs.pydantic.dev/latest/migration/#changes-to-config

File: python/kserve/kserve/protocol/rest/v2_datamodels.py [hunk 4]
```diff
@@ -183,7 +183,7 @@ class InferenceRequest(BaseModel):
           "outputs" : [ $request_output, ... ] #optional
         }
     """
-    id: Optional[str]
+    id: Optional[str] = None
     parameters: Optional[Parameters] = None
     inputs: List[RequestInput]
     outputs: Optional[List[RequestOutput]] = None
```
Comment (timothyjlaurent) [inline python/kserve/kserve/protocol/rest/v2_datamodels.py:189]: We need to replace Config with model_config if we want it to work in pydantic 2

File: python/kserve/kserve/protocol/rest/v2_datamodels.py (hunk 1 above)
Comment (timothyjlaurent) [reply python/kserve/kserve/protocol/rest/v2_datamodels.py:17]: If we do -- we should likely also likely add if/else for the the class definitions to switch from the nested Config class to model_config

File: python/kserve/kserve/protocol/rest/v2_datamodels.py (hunk 2 above)
Comment (timothyjlaurent) [reply python/kserve/kserve/protocol/rest/v2_datamodels.py:236]: This is to suppress this warning:

```
E   UserWarning: Field "model_name" has conflict with protected namespace "model_".
E
E   You may be able to resolve this warning by setting `model_config['protected_namespaces'] = ()`.
```

So it really ought to be an empty tuple.

**Comment by sivanantha321** [comment]
closing in favor of #3374

**Pull request closed without merging**